src/storage/src/controller.rs

```diff
@@ -1429,15 +1429,57 @@ mod persist_write_handles {
                                         )
-                                        .instrument(span)
-                                        .await
-                                        .expect("cannot append updates")
+                                        .instrument(span.clone())
+                                        .await;
+
+                                    // Indeterminate results can occur when persist is not certain
+                                    // whether the transaction has applied or not. We will attempt
+                                    // to suss this out by looking at the recent `upper`, and retrying
+                                    // if it is still appropriate, not retrying if it has advanced
+                                    // to `new_upper`, and panicking if it is anything else.
+                                    while let Err(indeterminate) = result {
+                                        tracing::warn!("Retrying indeterminate table write: {:?}", indeterminate);
+                                        write.fetch_recent_upper().await;
+                                        if write.upper() == &persist_upper {
+                                            // If the upper frontier is the prior frontier, the commit
+                                            // did not happen and we should retry it.
+                                            result =
+                                            write
+                                                .compare_and_append(
+                                                    updates.clone(),
+                                                    persist_upper.clone(),
+                                                    new_upper.clone(),
+                                                )
+                                                .instrument(span.clone())
+                                                .await;
+
+                                        } else if write.upper() == &new_upper {
+                                            // If the upper frontier is the new frontier, then because
+                                            // of mutual exclusion of writes, no other writer should be
+                                            // advancing the frontier to `new_upper`.
+                                            //
+                                            // TODO: This may succeed if `new_upper` is where we cut over
+                                            // to a new leader, who advanced tables to `new_upper` when it
+                                            // started. In that case, a success here will soon be followed
+                                            // by a failure on our next interaction with the catalog stash,
+                                            // but we would incorrectly think this committed and may serve
+                                            // results in the meantime.
```

Comment:
Maybe we should change the timestamp oracle to avoid this?

Reply:
Currently the timestamp oracle will make sure that every timestamp it uses is durable, but not necessarily every upper it uses. We could change this so the timestamp oracle makes sure the `upper` is durable. It would probably involve just adding a `step_forward` somewhere and would prevent this scenario.

Reply:
I think we can definitely solve the problem! I don't have a clear sense of the tidiest way to do that, though. But, worth chewing on if / as this lands.

Reply:
Ok, so I looked into it and I think we already do this so this is impossible. Let's say that `ts` is the upper bound saved in the stash.
- When a new leader starts up, it will perform empty writes at all tables at `ts` and advance all tables to `ts + 1`.
- The highest write a current leader could do is at `ts - 1` and advance all tables to `ts`. Before attempting to write at `ts` and advancing all tables to `ts + 1`, it must update the timestamp in the stash.

Reply:
Sweet. I may merge this with the comment in, and we can rip it out at the same time as we document that it is a reason why we have certain behavior elsewhere in the system / "please don't ever change it".